File: Old/Logistic Regression.py
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
from tqdm import tqdm
import seaborn as sns
import time

# Load MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Flatten images for Logistic Regression (28x28 -> 784)
train_images_flat = train_images.reshape(train_images.shape[0], -1) / 255.0
test_images_flat = test_images.reshape(test_images.shape[0], -1) / 255.0

# Initialize Logistic Regression with tuned settings
model_lr = LogisticRegression(
    max_iter=500,          # Lower max iterations to prevent excessive time
    solver='lbfgs',        # Optimized solver for multi-class classification
    n_jobs=-1,             # Use all CPU cores
    verbose=1              # Enable verbose output to track training progress
)

# Measure training time with progress tracking
print("Training Logistic Regression...")
start_time = time.time()
model_lr.fit(train_images_flat, train_labels)
training_time = time.time() - start_time
print(f"Training completed in {training_time:.2f} seconds.")

# Measure inference time
print("Running predictions...")
start_time = time.time()
lr_predictions = model_lr.predict(test_images_flat)
inference_time = time.time() - start_time
print(f"Inference completed in {inference_time:.2f} seconds.")

# Evaluate accuracy
lr_accuracy = accuracy_score(test_labels, lr_predictions)
print(f"Logistic Regression Accuracy: {lr_accuracy * 100:.2f}%")

# Confusion Matrix (to identify frequently misclassified digits)
cm = confusion_matrix(test_labels, lr_predictions)
plt.figure(figsize=(10, 8))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False)
plt.xlabel('Predicted')
plt.ylabel('True')
plt.title('Confusion Matrix for Logistic Regression')
plt.show()

# Classification Report (precision, recall, F1-score)
print("Classification Report:")
print(classification_report(test_labels, lr_predictions))

# Visualize misclassified digits for each class
misclassified = np.where(test_labels != lr_predictions)[0]
misclassified_by_class = {i: [] for i in range(10)}

# Group misclassified samples by true label
for idx in misclassified:
    true_label = test_labels[idx]
    misclassified_by_class[true_label].append(idx)

# Plot one misclassified example per class
plt.figure(figsize=(15, 8))
for i in range(10):
    if misclassified_by_class[i]:  # Check if there are misclassified samples for this class
        idx = misclassified_by_class[i][0]  # Take the first misclassified sample for this class
        plt.subplot(2, 5, i+1)
        plt.imshow(test_images[idx].reshape(28, 28), cmap='gray')
        plt.title(f"True: {test_labels[idx]}\nPred: {lr_predictions[idx]}")
    else:
        plt.subplot(2, 5, i+1)
        plt.text(0.5, 0.5, f"No misclassified\nfor {i}", ha='center', va='center')
        plt.axis('off')
plt.suptitle('Misclassified Digits by Logistic Regression (One Example per Class)')
plt.show()

# Calculate misclassified counts by digit
misclassified_counts = {i: 0 for i in range(10)}
for idx in misclassified:
    true_label = test_labels[idx]
    misclassified_counts[true_label] += 1

# Calculate successful classification percentage by digit
success_percentage = {i: 0 for i in range(10)}
for i in range(10):
    total_samples = np.sum(test_labels == i)
    correct_samples = total_samples - misclassified_counts[i]
    success_percentage[i] = (correct_samples / total_samples) * 100

# Print results
print(f"Logistic Regression Model Training Time: {training_time:.2f} seconds")
print(f"Logistic Regression Model Testing Time: {inference_time:.2f} seconds")
print(f"Logistic Regression Model Test Accuracy: {lr_accuracy * 100:.2f}%")
print("\nMisclassified counts by digit for Logistic Regression:")
for i in range(10):
    print(f"Digit {i}: {misclassified_counts[i]}")
print("\nSuccessful classification percentage for Logistic Regression by digit:")
for i in range(10):
    print(f"Digit {i}: {success_percentage[i]:.2f}%")

File: Old/Outputs/Logistic Regression/Logistic Regression.rs
Training Logistic Regression...
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
Training completed in 44.45 seconds.
Running predictions...
Inference completed in 0.03 seconds.
Logistic Regression Accuracy: 92.60%
2025-01-28 13:59:46.242 Python[79351:3758895] +[IMKClient subclass]: chose IMKClient_Modern
2025-01-28 13:59:46.242 Python[79351:3758895] +[IMKInputSession subclass]: chose IMKInputSession_Modern
Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.98      0.96       980
           1       0.96      0.98      0.97      1135
           2       0.93      0.90      0.91      1032
           3       0.90      0.91      0.91      1010
           4       0.94      0.94      0.94       982
           5       0.90      0.87      0.88       892
           6       0.94      0.95      0.95       958
           7       0.93      0.92      0.93      1028
           8       0.88      0.88      0.88       974
           9       0.91      0.92      0.91      1009

    accuracy                           0.93     10000
   macro avg       0.93      0.92      0.92     10000
weighted avg       0.93      0.93      0.93     10000

Logistic Regression Model Training Time: 44.45 seconds
Logistic Regression Model Testing Time: 0.03 seconds
Logistic Regression Model Test Accuracy: 92.60%

Misclassified counts by digit for Logistic Regression:
Digit 0: 22
Digit 1: 25
Digit 2: 101
Digit 3: 87
Digit 4: 60
Digit 5: 115
Digit 6: 47
Digit 7: 80
Digit 8: 119
Digit 9: 84

Successful classification percentage for Logistic Regression by digit:
Digit 0: 97.76%
Digit 1: 97.80%
Digit 2: 90.21%
Digit 3: 91.39%
Digit 4: 93.89%
Digit 5: 87.11%
Digit 6: 95.09%
Digit 7: 92.22%
Digit 8: 87.78%
Digit 9: 91.67%